src/sudoers/ast.rs

```diff
@@ -696,15 +700,26 @@ fn get_directive(
         "Host_Alias" => make(HostAlias(expect_nonterminal(stream)?)),
         "Cmnd_Alias" | "Cmd_Alias" => make(CmndAlias(expect_nonterminal(stream)?)),
         "Runas_Alias" => make(RunasAlias(expect_nonterminal(stream)?)),
-        "Defaults" => {
-            //HACK: this avoids having to add "Defaults@" etc as separate tokens; but relying
-            //on positional information during parsing is of course, cheating.
+        _ if keyword.starts_with("Defaults") => {
+            //HACK #1: no space is allowed between "Defaults" and '!>@:'. The below avoids having to
+            //add "Defaults!" etc as separate tokens; but relying on positional information during
+            //parsing is of course, cheating.
+            //HACK #2: '@' can be part of a username, so it will already have been parsed;
+            //an acceptable hostname is subset of an acceptable username, so that's actually OK.
+            //This resolves an ambiguity in the grammar similarly to how MetaOrTag does that.
+            const DEFAULTS_LEN: usize = "Defaults".len();
             let allow_scope_modifier = stream.get_pos().0 == begin_pos.0
-                && stream.get_pos().1 - begin_pos.1 == "Defaults".len();
+                && (stream.get_pos().1 - begin_pos.1 == DEFAULTS_LEN
+                    || keyword.len() > DEFAULTS_LEN);
 
             let scope = if allow_scope_modifier {
-                if is_syntax('@', stream)? {
-                    ConfigScope::Host(expect_nonterminal(stream)?)
+                if keyword[DEFAULTS_LEN..].starts_with('@') {
+                    let inner_stream = &mut CharStream::new_with_pos(
+                        &keyword[DEFAULTS_LEN + 1..],
+                        advance(begin_pos, DEFAULTS_LEN + 1),
+                    );
```

Comment:
Maybe we could have a method on `CharStream` to advance a given amount of positions and then use `CharStream::new_with_pos(&keyword, begin_pos)` followed by calling this method with `DEFAULTS_LEN + 1` as argument? The code you have right now is fine too though.Maybe we could have a method on `CharStream` to advance a given amount of positions and then use `CharStream::new_with_pos(&keyword, begin_pos)` followed by calling this method with `DEFAULTS_LEN + 1` as argument? The code you have right now is fine too though.

Reply:
I like this idea. I'd prefer to handle that in a separate PR!